.github/workflows/helpers/pull-request-utils.js

```diff
@@ -0,0 +1,331 @@
+    @returns {Array} - filtered commit diff, including only the files that are part of the PR diff
+  */
+  async #filterCommitDiff(commitDiff = [], prDiff = []) {
+    return commitDiff.filter((file) => prDiff.includes(file))
+  }
+
+  /*
+    Builds the diff for the pull request, including both the changes in the pull request and the changes in non-scanned commits
+    @returns {string} - json string representation of the pull request diff and the diff for non-scanned commits
+  */
+  async buildDiff() {
+    const { data } = await this.github.rest.pulls.listFiles({
+      owner: this.owner,
+      repo: this.repo,
+      pull_number: this.pullRequestNumber,
+      per_page: resultSize,
+    })
+
+    const pullRequestDiff = await this.#getDiffForFiles(data)
+
+    let nonScannedCommitsDiff = {}
+
+    // The "synchronize" event implies that new commit are pushed after the pull request was opened
+    if (
+      Object.keys(pullRequestDiff).length != 0 &&
+      this.pullRequestEvent === synchronizeEvent
+    ) {
+      // Retrieves list of commits that have not been scanned by the PR check
+      const nonScannedCommits = await this.#getNonScannedCommits()
+      for (const commit of nonScannedCommits) {
+        const { data } = await this.github.rest.repos.getCommit({
+          owner: this.owner,
+          repo: this.repo,
+          ref: commit,
+        })
+
+        const commitDiff = await this.#getDiffForFiles(data.files)
+        const files = Object.keys(commitDiff)
+        for (const file of files) {
+          // Consider scenario where the changes made to a file in the initial commit are completely undone by subsequent commits
+          // In such cases, the modifications from the initial commit should not be taken into account
+          // If the changes were entirely removed, there should be no entry for the file in the pullRequestStats
+          const filePRDiff = pullRequestDiff[file]
+          if (!filePRDiff) {
+            continue
+          }
+
+          // Consider scenario where changes made in the commit were partially removed or modified by subsequent commits
+          // In such cases, include only those commit changes that are part of the pullRequestStats object
+          // This ensures that only the changes that are reflected in the pull request are considered
+          const changes = await this.#filterCommitDiff(commitDiff[file], filePRDiff)
+
+          if (changes.length !== 0) {
+            // Check if nonScannedCommitsDiff[file] exists, if not assign an empty array to it
+            nonScannedCommitsDiff[file] = nonScannedCommitsDiff[file] || []
+            // Combine the existing nonScannedCommitsDiff[file] array with the commit changes
+            // Remove any duplicate elements using the Set data structure
+            nonScannedCommitsDiff[file] = [
+              ...new Set([...nonScannedCommitsDiff[file], ...changes]),
+            ]
+          }
+        }
+      }
+    }
```

Comment:
Do you think moving this code to separate function will help in code readibility? function like `RetrieveNonScannedCommit`?

Reply:
https://github.com/prebid/prebid-server/pull/2907/commits/9d06e6b58b99384e0a62678c08efeb9e35fc79d0 

makes above sense